staging/src/k8s.io/apiserver/pkg/storage/etcd3/lease_manager.go

```diff
@@ -0,0 +1,207 @@
+	}
+}
+
+// setLeaseReuseDurationSeconds is used for testing purpose. It is used to
+// reduce the extra lease duration to avoid unnecessary timeout in testing.
+func (l *leaseManager) setLeaseReuseDurationSeconds(duration int64) {
+	l.leaseMu.Lock()
+	defer l.leaseMu.Unlock()
+	l.leaseReuseDurationSeconds = duration
+}
+
+// GetLease returns a lease based on requested ttl: if a cached lease can be
+// reused, reuse it; otherwise request a new one from etcd.
+func (l *leaseManager) GetLease(ctx context.Context, ttl int64) (clientv3.LeaseID, error) {
+	now := time.Now().Unix()
+	id := l.getLeaseFromCache(now, ttl)()
+	if id > clientv3.LeaseID(0) {
+		return id, nil
+	}
+	id, err := l.grantLeaseFromEtcd(ctx, ttl)
+	l.storeLease(id, now, ttl)
+	return id, err
+}
+
+// grantLeaseFromEtcd requests a lease for certain ttl from etcd.
+func (l *leaseManager) grantLeaseFromEtcd(ctx context.Context, ttl int64) (clientv3.LeaseID, error) {
+	reuseDurationSeconds := l.getReuseDurationSeconds(ttl)
+	lcr, err := l.client.Lease.Grant(ctx, ttl+reuseDurationSeconds)
+	if err != nil {
+		return clientv3.LeaseID(-1), err
+	}
+	return lcr.ID, nil
+}
+
+// getLeaseFromCache checks whether the requested lease with certain ttl can
+// reuse any of the cached leases: if yes, return the lease id; otherwise,
+// return error indicated by leaseID=-1.
+func (l *leaseManager) getLeaseFromCache(now, ttl int64) func() clientv3.LeaseID {
+	l.leaseMu.Lock()
+	defer l.leaseMu.Unlock()
+	for ti, id := range l.leaseCache {
+		if ti <= now { // delete expired leases as checking
+			delete(l.leaseCache, ti)
+			delete(l.leaseLastUsedTime, ti)
+		} else {
+			if l.isReusable(ti-now, ttl) { // check if the lease is reusable
+				if id > clientv3.LeaseID(0) { // if reusable and the lease id is available, return it
+					l.leaseLastUsedTime[ti] = now
+					return func() clientv3.LeaseID { return id }
+				}
+				// if reusable but the lease id is not available,
+				// register a chan for notification and wait for the lease id
+				c := make(chan clientv3.LeaseID, 1)
+				if _, ok := l.leaseNotification[ti]; !ok {
+					l.leaseNotification[ti] = make([]chan clientv3.LeaseID, 0)
+				}
+				l.leaseNotification[ti] = append(l.leaseNotification[ti], c)
+				return func() clientv3.LeaseID {
+					id = <-c
+					close(c)
+					return id
+				}
+			}
+		}
+	}
+	return func() clientv3.LeaseID { return clientv3.LeaseID(-1) }
+}
+
+// storeLease stores lease to its cache, to be reused in future.
+func (l *leaseManager) storeLease(id clientv3.LeaseID, now, ttl int64) {
+	l.leaseMu.Lock()
+	defer l.leaseMu.Unlock()
+	// Firstly, notify all the listeners the lease id
+	reuseDurationSeconds := l.getReuseDurationSeconds(ttl)
+	ti := now + ttl + reuseDurationSeconds
+	for _, c := range l.leaseNotification[ti] {
+		c <- id
+	}
+	delete(l.leaseNotification, ti)
+	// Secondly, update the lease cache
+	if id <= clientv3.LeaseID(0) { // if returned id is invalid, clean the maps
```

Comment:
remove comments describing what the code does when it is obvious (there are multiple places like this one)